README.md

```diff
@@ -1,25 +1,151 @@
+-->
+
+
+<!--
+## Computing Jacobians
+
+-- TODO: we can have vector/matrix/tensor codomains, but not pair codomains
+-- until #68 is done;
+-- perhaps a vector codomain example, with a 1000x3 Jacobian, would make sense?
+
+Now let's consider a function from 'R^n` to `R^m'.  We don't want the gradient, but instead the Jacobian.
+```hs
+-- A function that goes from `R^3` to `R^2`.
+foo :: RealFloat a => (a,a,a) -> (a,a)
+foo (x,y,z) =
+  let w = x * sin y
+  in (atan2 z w, z * w)
+```
+TODO: show how the 2x3 Jacobian emerges from here
+-->
+
+
+# Forall shapes and sizes
+
+An additional feature of this library is a type system for tensor shape arithmetic.  Look at this beauty.
+```hs
+convMnistTwoS
+  kh@MkSN kw@MkSN h@MkSN w@MkSN c_in@MkSN c_out@MkSN _n_hidden@MkSN batch_size@MkSN
+    -- integer parameters denoting basic dimensions, with some notational noise
+  input              -- input images, shape (batch_size, c_in, h, w)
+  (ker1, bias1)      -- layer1 kernel, shape (c_out, c_in, kh+1, kw+1); and bias, shape (c_out)
+  (ker2, bias2)      -- layer2 kernel, shape (c_out, c_out, kh+1, kw+1); and bias, shape (c_out)
+  ( weigthsDense     -- dense layer weights,
+                     -- shape (n_hidden, c_out * ((h+kh)/2 + kh)/2, ((w+kw)/2 + kw)/2)
+  , biasesDense )    -- dense layer biases, shape (n_hidden)
+  ( weigthsReadout   -- readout layer weights, shape (10, n_hidden)
+  , biasesReadout )  -- readout layer biases (10)
+  =                  -- -> output classification, shape (10, batch_size)
+  let t1 = convMnistLayerS kh kw
+                           h w
+                           c_in c_out batch_size
+                           ker1 (constant input) bias1
+      t2 = convMnistLayerS kh kw
+                           (MkSN @((h + kh) `Div` 2)) (MkSN @((w + kw) `Div` 2))
+                           c_out c_out batch_size
+                           ker2 t1 bias2
+      m1 = mapOuterS reshapeS t2
+      m2 = transpose2S m1
+      denseLayer = weigthsDense <>$ m2 + asColumnS biasesDense
+      denseRelu = relu denseLayer
+  in weigthsReadout <>$ denseRelu + asColumnS biasesReadout
+```
+But we don't just want the shapes in comments and in runtime expressions;
+we want them in the type system to be checked at compile time:
+```hs
+convMnistTwoS
+  :: forall kh kw h w c_in c_out n_hidden batch_size d r.
+     ( 1 <= kh             -- kernel height is large enough
+     , 1 <= kw             -- kernel width is large enough
+     , ADModeAndNum d r )  -- differentiation mode and numeric type are known to the engine
+  => -- The two boilerplate lines below tie type parameters to the corresponding
+     -- value parameters (built with MkSN) denoting basic dimensions.
+     StaticNat kh -> StaticNat kw -> StaticNat h -> StaticNat w
+  -> StaticNat c_in -> StaticNat c_out -> StaticNat n_hidden -> StaticNat batch_size
+  -> OS.Array '[batch_size, c_in, h, w] r
+  -> ( ADVal d (OS.Array '[c_out, c_in, kh + 1, kw + 1] r)
```

Comment:
How did you know? `git stash list -i -GADArray` turned out two stashes with exactly this type synonym. :) I eventually haven't asked for feedback on that one in the PR, because I thought this sugar obscures the relation between `OS.Array l r` and `ADVal d (OS.Array l' r)` on the consecutive lines above, which is that the latter is the former wrapped in "this value will be differentiated".

In any case, if @simonpj insists, we are going to switch to `ArrayLike arr => arr l r -> ...` anyway. Oh, wait, in this very example (and so probably in most of the convolutional nn code) we can't, because we explicitly differentiate a training data batch that is not to be differentiated against and all the other array parameters that are to be trained. So we'd need to mix `ArrayLike arr`, `ADVal` and, if we so choose, also `ADArray`. What do you think?